enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,365 @@
+decision should be made.
+
+#### Hypershift
+
+Hypershift does not use `ControlPlaneMachineSet` to manage the control-planes
+for hosted clusters, as those are not running on their own virtual machines.
+
+Because of this, no special handling should be required.
+
+### Risks and Mitigations
+
+#### Risk 1: Higher costs when scaling up
+
+Scaling up will increase costs for running the cluster, as bigger nodes will
+incur higher costs from the cloud provider.
+
+#### Risk 2: Cluster instability due to scaling down
+
+Scaling down might also impact the clusters ability to accommodate a workload if
+resource usage was uncharacteristically low for the detection period of scaling.
+This turns into an issue, if the cluster still would have required the bigger
+control plane nodes.
+
+To mitigate this issue the scale down behavior should be disabled by default and
+require an explicit opt-in by the cluster administrator.
+
+#### Risk 3: Control plane node churn
+
+Automatically scaling up and down control planes could impact cluster
+reliability and availability, if performed too often.
+
+As such the additional logic must ensure that there is a grace period between
+changes to the machine size, during which no more scaling will be performed.
+
+### Drawbacks
+
+Some users might not want this feature and rather opt to scale control planes
+manually.
+
+## Design Details
+
+### Control Plane MachinSet operator configuration
+
+As automatic scaling will require adjustments to certain aspects of its logic,
+the implementation should implement a way for administrators of an OpenShift
+cluster to configure the specifics of how and when scaling will occur.
+
+This configuration will be performed using a new custom resource
+`ControlPlaneMachineSetAutoscaling`.
+
+The **required** configurations will include the following properties:
+
+- `instancesSizes`: a YAML object matching the [provider specific
+  configuration](https://docs.openshift.com/container-platform/4.13/machine_management/control_plane_machine_management/cpmso-configuration.html)
+  for the used cloud platform with values that must be adjusted. The keys must
+  be rooted under the `providerSpec.value` field: So for `Azure` cluster this
+  should set the `vmSize`, for `GCP` the `machineType`, for `AWS` the
+  `instanceType` and for `vSphere` `memoryMiB`, `numCPUs` and `diskGiB`. It also
+  has to include a `weight` property to specify the order in which scaling
+  should occur.
+
+The **optional** configurations will include the following properties:
+
+- `syncPeriod`: how often will the operator check resource usage to see if scaling
+  should be performed.
+- `gracePeriod`: how long after a successful scaling should no more scaling be
+  performed, even if the thresholds would scale the control planes again.
+- `scaleDownEnabled`: can enable or disable scaling down.
+- `scaleUpEnabled`: can enable or disable scaling up.
```

Comment:
[The horizontal pod autoscaler][1] has:

```yaml
    scaleUp:
      selectPolicy: Disabled
```

which sets up a nice [discriminated union][2] pattern that allows convenient extensibility if later maintainers think of new schemas for either scale-up or scale-down.

[1]: https://docs.openshift.com/container-platform/4.13/nodes/pods/nodes-pods-autoscaling.html#nodes-pods-autoscaling-policies_nodes-pods-autoscaling
[2]: https://github.com/openshift/enhancements/blob/da2ef4e0702751c3f5f7faa393282b5dd823195e/dev-guide/api-conventions.md#discriminated-unions

Reply:
Thanks for the great link - I've also used the same pattern for the cloud provider specific machine configuration now.